pkg/rbd/controllerserver.go

```diff
@@ -162,3 +190,223 @@ func (cs *controllerServer) ControllerUnpublishVolume(ctx context.Context, req *
+		glog.Warningf("invalid create snapshot req: %v", req)
+		return nil, err
+	}
+
+	// Check sanity of request Snapshot Name, Source Volume Id
+	if len(req.Name) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Snapshot Name cannot be empty")
+	}
+	if len(req.SourceVolumeId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Source Volume ID cannot be empty")
+	}
+
+	// Need to check for already existing snapshot name, and if found
+	// check for the requested source volume id and already allocated source volume id
+	if exSnap, err := getRBDSnapshotByName(req.GetName()); err == nil {
+		if req.SourceVolumeId == exSnap.SourceVolumeID {
+			return &csi.CreateSnapshotResponse{
+				Snapshot: &csi.Snapshot{
+					SizeBytes:      exSnap.SizeBytes,
+					Id:             exSnap.SnapID,
+					SourceVolumeId: exSnap.SourceVolumeID,
+					CreatedAt:      exSnap.CreatedAt,
+					Status: &csi.SnapshotStatus{
+						Type: csi.SnapshotStatus_READY,
+					},
+				},
+			}, nil
+		}
+		return nil, status.Error(codes.AlreadyExists, fmt.Sprintf("Snapshot with the same name: %s but with different source volume id already exist", req.GetName()))
+	}
+
+	rbdSnap, err := getRBDSnapshotOptions(req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	// Generating Snapshot Name and Snapshot ID, as according to CSI spec they MUST be different
+	snapName := req.GetName()
+	uniqueID := uuid.NewUUID().String()
+	rbdVolume, err := getRBDVolumeByID(req.GetSourceVolumeId())
+	if err != nil {
+		return nil, status.Error(codes.NotFound, fmt.Sprintf("Source Volume ID %s cannot found", req.GetSourceVolumeId()))
+	}
+	if !hasSnapshotFeature(rbdVolume.ImageFeatures) {
+		return nil, fmt.Errorf("Volume(%s) has not snapshot feature(layering)", req.GetSourceVolumeId())
+	}
+
+	rbdSnap.VolName = rbdVolume.VolName
+	rbdSnap.SnapName = snapName
+	snapshotID := "csi-rbd-" + rbdVolume.VolName + "-snap-" + uniqueID
+	rbdSnap.SnapID = snapshotID
+	rbdSnap.SourceVolumeID = req.GetSourceVolumeId()
+	rbdSnap.SizeBytes = rbdVolume.VolSize
+
+	err = createSnapshot(rbdSnap, rbdSnap.AdminId, req.GetCreateSnapshotSecrets())
+	// if we already have the snapshot, return the snapshot
+	if err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
+				if status.ExitStatus() == int(syscall.EEXIST) {
+					glog.Warningf("Snapshot with the same name: %s, we return this.", req.GetName())
+				} else {
+					glog.Warningf("failed to create snapshot: %v", err)
+					return nil, err
+				}
+			} else {
+				glog.Warningf("failed to create snapshot: %v", err)
+				return nil, err
+			}
+		} else {
+			glog.Warningf("failed to create snapshot: %v", err)
+			return nil, err
+		}
+	} else {
+		glog.V(4).Infof("create snapshot %s", snapName)
+		err = protectSnapshot(rbdSnap, rbdSnap.AdminId, req.GetCreateSnapshotSecrets())
```

Comment:
thanks @dillaman 

@sngchlko do you have a chance to consider this?

Reply:
@dillaman Thanks for the review. Good point!!
But I think only supporting snapshot with Mimic is not good.
In my case, I have to use the luminous version as I installed the ceph with a rook project which supports only luminous now.
And people who want to use snapshot feature in CSI must upgrade to Mimic.
How about using protect operation according to version?

Reply:
@sngchlko Like I said, the "protect" operation is only relevant for cloning images (using older versions of Ceph). Creating a snapshot only requires the "rbd snap create" command.

Reply:
@dillaman You are right. But I think we need protecting snapshot for creating a new image from the snapshot. CSI spec seems to require creating new volume at the time of restoring the snapshot. AFAIK, cloning the snapshot requires protecting the snapshot.

Reply:
@sngchlko I'm not familiar w/ the CSI spec, can you provide a link? I see [1] but it states revert isn't part of it. I would really like to avoid going down the tangled code of the OpenStack world due to how snapshots and clones needed to be handled before, which is why we added clone v2. Since this is all new functionality, it would be great if we could start off on a clean footing and avoid having to handle things like renaming protected snapshots "XYZ_deleteme" and images as "XYZ_deleteme" if they have a linked clone.

[1] https://docs.google.com/document/d/1oXVuDTNhXerr1UJ48tQrxF9J6vbm2LHVVC3z28655Qo/edit#heading=h.aqok97b5hs7t

Reply:
@dillaman Sure, you can find `VolumeContentSource` for restoring the snapshot in [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume). If `VolumeContentSource` has something, it creates the new volume from the snapshot.
I think I can't determine about supporting snapshot only Mimic.
@rootfs, @gman0 What do you think about this?

Reply:
@sngchlko Thanks. If the only use-case for snapshots right now is to restore and if restore requires creating new volumes (instead of rolling back), perhaps the short-term (pre-Mimic) thing to do would be to protect the snapshot upon create from snapshot, flatten the new volume, and unprotect the snapshot. Mimic and later clients can just directly clone the snapshot into a new image. --- or just only support creating images from a snapshot if you are using Mimic or later and enjoy the simpler CSI code that doesn't need to manage snapshot lifecycles.

Reply:
@dillaman I fully understand your concerns. I created https://github.com/ceph/ceph-csi/issues/70. Thanks!